### ISSUE TITLE
Pending writes (flush) not flushed when client invokes socket.shutdownOutput() with out closing the socket.

### ISSUE THREAD

**Issue by Ramana-Muthyala** [netty/netty#13072]
### Steps to reproduce
1. Connect to server
2. Write data
3. invoke socket.shutdownOutput()
4. wait for response  --  netty does not flush pending writes

### Minimal yet complete reproducer code (or URL to code)
Please see below comment for scenario and reproducing steps.

**Comment by Ramana-Muthyala** [comment]
Was expecting a bad request for invalid HTTP request.
Writes are flushed for valid requests.

**Comment by Ramana-Muthyala** [comment]
Sorry for closing the bug. Was confused a little. Reopening it now. Providing information below:

Scenario:
There could be cases where client invokes socket.shutdownOutput() and then attempts to read the response.
In such cases:
1. socket.shutdownOutput() triggers channelReadComplete in netty.
2. After read complete, netty flushes pending writes and then closes the socket.
3. There could be scenarios where due to socket output buffer full, the flush may not have been honored.
        In such cases, the write has to wait for the next event loop cycle but by then the socket is already closed.
4. As a result, the client does not receive response.

This is not a common use case. It is a border scenario and may occur rarely.
To simulate this, I modified the source code of io.netty.channel.socket.nio.NioSocketChannel#doWrite
I prevented write for subsequent 2 event loop cycles as given below: 
Added a field blockFlushCounter to simulate the border case.

ByteBuffer buffer = nioBuffers[0];
                    int attemptedBytes = buffer.remaining();
                    int localWrittenBytes = 0;
                    if(blockFlushCounter < 2) {
                        // simulating write failure (socket write buffer may be full)
                        // simulating failure for 2 times
                        blockFlushCounter++;
                    } else {
                        localWrittenBytes = ch.write(buffer);
                    }
                    if (localWrittenBytes <= 0) {
                        incompleteWrite(true);
                        return;
                    }


This bug could occur under heavy load where output socket buffer is full. 
Please clarify in case I am wrong or fix it if what I reported is correct.

Thank you.

**Comment by chrisvest** [comment]
Writes are asynchronous in Netty. If you want to do something _after_ a write has completed, you need to attach a listener to the Future we return from write or writeAndFlush.

**Comment by Ramana-Muthyala** [comment]
> Writes are asynchronous in Netty. If you want to do something _after_ a write has completed, you need to attach a listener to the Future we return from write or writeAndFlush.

Yes. writes are asynchronous. I actually meant flush is not done in the above mentioned scenario. Please see the code excerpt. It is related to flushing. 
Thank you for responding.

**Comment by normanmaurer** [comment]
Like @chrisvest said... you will need to flush by yourself before calling `shutdownOutput`

**Comment by Ramana-Muthyala** [comment]
Hi Norman
I think you haven't seen the comment fully.
Pasting it here:

Scenario:
There could be cases where client invokes socket.shutdownOutput() and then attempts to read the response.
In such cases:

socket.shutdownOutput() triggers channelReadComplete in netty.
After read complete, netty flushes pending writes and then closes the socket.
There could be scenarios where due to socket output buffer full, the flush may not have been honored.
In such cases, the write has to wait for the next event loop cycle but by then the socket is already closed.
As a result, the client does not receive response.
This is not a common use case. It is a border scenario and may occur rarely.
To simulate this, I modified the source code of io.netty.channel.socket.nio.NioSocketChannel#doWrite
I prevented write for subsequent 2 event loop cycles as given below:
Added a field blockFlushCounter to simulate the border case.

ByteBuffer buffer = nioBuffers[0];
int attemptedBytes = buffer.remaining();
int localWrittenBytes = 0;
if(blockFlushCounter < 2) {
// simulating write failure (socket write buffer may be full)
// simulating failure for 2 times
blockFlushCounter++;
} else {
localWrittenBytes = ch.write(buffer);
}
if (localWrittenBytes <= 0) {
incompleteWrite(true);
return;
}

This bug could occur under heavy load where output socket buffer is full.
